deps/rabbit/src/rabbit_classic_queue.erl

```diff
@@ -63,14 +64,61 @@
          send_drained_credit_api_v1/4,
          send_credit_reply/7]).
 
+-export([validate_policy/1]).
+
+-rabbit_boot_step(
+   {?MODULE,
+    [{description, "Deprecated queue-master-locator support."
+      "Use queue-leader-locator instead."},
+     {mfa, {rabbit_registry, register,
+            [policy_validator, <<"queue-master-locator">>, ?MODULE]}},
+     {mfa, {rabbit_registry, register,
+            [operator_policy_validator, <<"queue-master-locator">>, ?MODULE]}},
+     {requires, rabbit_registry},
+     {enables, recovery}]}).
+
+validate_policy(Args) ->
+    Strategy = proplists:get_value(<<"queue-master-locator">>, Args, unknown),
+    case rabbit_queue_location:master_locator_permitted() of
+        true ->
+            case lists:member(Strategy, rabbit_queue_location:queue_leader_locators()) of
+                true -> ok;
+                false -> {error, "~tp is not a valid master locator", [Strategy]}
+            end;
+        false ->
+            {error, "use of deprecated queue-master-locator argument is not permitted", []}
+    end.
+
 -spec is_enabled() -> boolean().
 is_enabled() -> true.
 
 -spec is_compatible(boolean(), boolean(), boolean()) -> boolean().
 is_compatible(_, _, _) ->
     true.
 
+validate_arguments(Args) ->
+    case lists:keyfind(<<"x-queue-master-locator">>, 1, Args) of
+        false ->
+            ok;
+        _ ->
+            case rabbit_queue_location:master_locator_permitted() of
+                true ->
+                    ok;
+                false ->
+                    error
+            end
+    end.
+
 declare(Q, Node) when ?amqqueue_is_classic(Q) ->
+    case validate_arguments(amqqueue:get_arguments(Q)) of
+        ok -> do_declare(Q, Node);
+        error ->
+            Warning = rabbit_deprecated_features:get_warning(
+                        queue_master_locator),
```

Comment:
It's better to return this warning and `protocol_error` directly in `validate_arguments/1` since right now `queue_master_locator` validation leaks out from function `validate_arguments/1` into `declare/2`. In other words, it's better when function `declare/2` doesn't know about `queue_master_locator`.